### ISSUE TITLE
Should GIDs be used in the input and output h5 files?

### ISSUE THREAD

**Issue by pgleeson** [AllenInstitute/sonata#46]
While GIDs are a useful & quick way to get data out of a running simulation, there are some reasons it may not be best to use these in the input and output formats:

1) The network is specified in terms of populations and node_ids, and if a user gets this specification and the result files using GIDs, they have to convert these back into population/node_ids before processing the results.
2) This mapping can optionally be included in a h5 file, but most examples don't include them.
3) The mapping itself may be implementation dependent: are the gids assigned per population in alphabetical order? This should be entirely up to the simulator.
4) If the user does want to reconstruct a missing GID list for a simulation, they will probably have to load the network into the simulator to get it out, which may take a while for a big network. It also presupposes that the user has access to the right version of the script used to generate that network. 
5) If the results files become dissociated from the exact simulation description for any reason they become almost unusable: the GIDs have no intrinsic meaning, it would be better at least if they contained population/node_ids.
6) Use of GID in input files presupposes these are known before the simulation. However in this example: https://github.com/kaeldai/sonata/tree/feature/300_intfire/examples/300_intfire/inputs there are 2 input sets to 2 populations, which both have GIDs 0-89...
7) If the network is reconfigured and one population (not the one stimulated) is removed/resized this could change all the GIDs, meaning existing input.h5 files need to be regenerated.

In short (IMHO...) GIDs are temporary ids generated during the simulation, and if input/output files are to refer to exact nodes in the network, they should ideally use the unique population/node_ids as used in the network specification.

**Comment by hernando** [comment]
Good that you bring up this point, because I was struggling to see how we can provide a convenient implementation of the GID -> population, node id mappings.

The GID -> population, node id mapping is indeed inconvenient, but having in mind that there are tools being used at the moment that need to be migrated to SONATA and that they are GID based, having GIDs in SONATA provides a smoother transitions path until SONATA is natively supported.

The mapping is not supposed to be optional, it's absolutely necessary to have a correct interpretation of input and ouput files in the context of a full network. The fact that it's missing from the examples is due undocumented simplifications and assumptions (e.g. GIDs map to node ids in the first population). 

In an early proposal we discussed not having node groups at all and making populations single homogeneous collections of nodes with implicitly assigned node ids based on the position of the node in the population (no node_id, node_group_id and node_index datasets). This alternative makes the GID -> population, node id straightforward for a given circuit given an explicit enumeration of the population involved, however it was rejected for various reasons. In this alternative output files could use GIDs without no problem as they are always related to a circuit which is immutable from their point of view. For input files, using the same file with different variations of a circuit would still be problematic.

Using (population id, node id) pairs as GIDs could be considered as long as population IDs are numbers and a supplementary table describes the mapping between population ids and names (unless this can be inferred implicitly from some other information). Using strings as the population IDs is a bad idea as it will make GID sets orders of magnitude bigger. In this case, some short of mapping will still need to be implemented during the transition period.

**Comment by hernando** [comment]
I expect @pgleeson's alternative to be adding an additional group below spikes with the population name. For consistency the same thing should be done for compartment (*aka* frame oriented node element) reports. This one looks potentially simpler, but requires writing to multiple datasets when reporting. I don't feel qualified to evaluate whether this is a potential problem or not for simulators. From the reading side I guess it's fine as we are not even considering populations at the moment.

In fact, having to choose a population before accessing any sort of data and renaming GIDs to NodeIDs may indeed be a better solution for us to fully support SONATA in our existing software rather than having datasets that mix different populations. I'm even starting to be more keen on this solution than what I just wrote as it seems to require fewer changes in our APIs and implementations, so I may write a proposal for this variation today as well.

**Comment by hernando** [comment]
I have create the other proposal. For convenience I'm adding both links here:
* [proposal 1](https://github.com/AllenInstitute/sonata/compare/master...hernando:removing_gids): adding extra datasets for the population ids
* [proposal 2](https://github.com/AllenInstitute/sonata/compare/master...hernando:removing_gids2): defining the population as an upper level group

**Comment by pgleeson** [comment]
Thanks @hernando. I certainly prefer proposal 2. I need to look through the documentation to see what this would entail, but as a general point I think the final spec should make it clear that GIDs are a purely optional attribute and all  sim output datasets should stand alone and be usable if the GID info is ignored.

**Comment by hernando** [comment]
I've created #73 with the agreed solution.
